### PR TITLE
Specific process for geometries with Z or M coordinate with SpatiaLite dialect

### PR DESCRIPTION
**Pull request by sdp5**
### Description

Though most of the dialects support  geom_type ending with Z in SQL queries, it seems not working with SpatiaLite. 
Hence replacing Z with a blank in SQL queries.

Added a test to cover this behaviour.

### Checklist

This pull request is:

- [x] A short code fix
	- Fixes: #505 

@adrien-berchet Please take an initial look, thanks!

### CODE REVIEW

**Comment by sdp5** [comment]
Agree, getting your point @adrien-berchet and thank you for setting up the tests. 👍🏻 

Allow me some more time testing, will update the PR shortly.

File: geoalchemy2/types/dialects/sqlite.py [hunk 1]
```diff
@@ -13,9 +13,16 @@ def bind_processor_process(spatial_type, bindvalue):
         else:
             return "SRID=%d;%s" % (bindvalue.srid, bindvalue.data)
     elif isinstance(bindvalue, WKBElement):
+        if bindvalue.srid == -1:
+            bindvalue.srid = spatial_type.srid
```
[the remark this answers is not in the document]
Comment (sdp5) [reply geoalchemy2/types/dialects/sqlite.py:35]: thank you for retaining them in the form of `forced_srid` ... I guess they are getting baked in.

```
'SRID=-1;LINESTRING Z (0 0 0, 1 1 1)'

0x0001FFFFFFFF0000000000 ...
```

```
'SRID=4326;LINESTRING Z (0 0 0, 1 1 1)'

0x0001E610000000000000000 ...
```

**Comment by sdp5** [comment]
> especially the M cases, like POINT M or POINT ZM, it should be added in the test

I guess support for 'M' is [coming in](https://github.com/shapely/shapely/issues/1648) shapely.

```
>>> from shapely import wkt
>>> g = wkt.loads('POINT M (0 0 5)')
>>> g
<POINT Z (0 0 5)>
```

_Perhaps 'M'  and 'ZM' support could be put on the TODO-list?_

**Comment by adrien-berchet** [comment]
> > especially the M cases, like POINT M or POINT ZM, it should be added in the test
> 
> I guess support for 'M' is [coming in](https://github.com/shapely/shapely/issues/1648) shapely.
> 
> ```
> >>> from shapely import wkt
> >>> g = wkt.loads('POINT M (0 0 5)')
> >>> g
> <POINT Z (0 0 5)>
> ```
> 
> _Perhaps 'M' and 'ZM' support could be put on the TODO-list?_
> 
> Plus with 'ZM' ... shapely returns `ValueError: The ordinate (last) dimension should be 2 or 3, got 4`!

Hmmm ok
Well, we can still process the WKT strings for M and ZM cases, even though Shapely does not support it yet. So it will still fix the cases with raw strings and WKTElements, only the WKBElements will fail.

File: tests/test_functional_sqlite.py [hunk 2]
```diff
@@ -129,6 +129,41 @@ def test_explicit_schema(self, conn):
         # Drop the table
         t.drop(bind=conn)
 
+    def test_3d_geometry(self, conn, metadata):
```
[the remark this answers is not in the document]
Comment (sdp5) [reply tests/test_functional_sqlite.py]: resolving thanks.

File: geoalchemy2/types/dialects/sqlite.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sdp5) [reply geoalchemy2/types/dialects/sqlite.py]: resolving thanks.

**Comment by adrien-berchet** [comment]
I'm not able to reproduce the CI failure locally, even in the test Docker container :thinking: 
We'll have to investigate that.

**Comment by sdp5** [comment]
ran some tests quickly with `postgis/postgis:15-3.3` image.
and found a few tests failing for `postgresql` dialect, `MULTIPOINT` geom_type.

```
FAILED tests/test_functional.py::TestInsertionCore::test_insert_all_geom_types[postgresql-Multi Point ZM]
AssertionError: assert 'MULTIPOINTZM((1234),(5678))' == 'MULTIPOINTZM(1234,5678)'
```

we can fix that by adding following in `test_insert_all_geom_types` 
```

if dialect_name == 'postgresql' and 'MULTIPOINT' in geom_type:
      # formats wkt to wrap tuple elements in brackets,
      # for example '(1 2, 3 4)' to '((1 2), (3 4))'.
      wkt = "({})".format(
           ", ".join(map(lambda x: '({})'.format(x.strip()), wkt[1:-1].split(',')))
      )
```

Can we use this image in `test_and_publish.yml`?

**Comment by sdp5** [comment]
> The only nice thing to have would be to update the doc to specify which cases are supported and which are not.

@adrien-berchet for docs, should we make it a part of `core_tutorial.rst#Insertions` ?

**Comment by adrien-berchet** [comment]
> ran some tests quickly with `postgis/postgis:15-3.3` image. and found a few tests failing for `postgresql` dialect, `MULTIPOINT` geom_type.
> 
> ```
> FAILED tests/test_functional.py::TestInsertionCore::test_insert_all_geom_types[postgresql-Multi Point ZM]
> AssertionError: assert 'MULTIPOINTZM((1234),(5678))' == 'MULTIPOINTZM(1234,5678)'
> ```
> 
> we can fix that by adding following in `test_insert_all_geom_types`
> 
> ```
> 
> if dialect_name == 'postgresql' and 'MULTIPOINT' in geom_type:
>       # formats wkt to wrap tuple elements in brackets,
>       # for example '(1 2, 3 4)' to '((1 2), (3 4))'.
>       wkt = "({})".format(
>            ", ".join(map(lambda x: '({})'.format(x.strip()), wkt[1:-1].split(',')))
>       )
> ```

I'm surprised they changed this behavior, that's quite annoying.

> Can we use [this image](https://github.com/postgis/docker-postgis) in `test_and_publish.yml`?

Yeah, let's go for a newer version if it does not break anything (except the `MULTIPOINT` thing ofc).

Thanks for this!

**Comment by adrien-berchet** [comment]
> > The only nice thing to have would be to update the doc to specify which cases are supported and which are not.
> 
> @adrien-berchet for docs, should we make it a part of `core_tutorial.rst#Insertions` ?

Yeah that looks relevant.

**Comment by adrien-berchet** [comment]
For the CI I see that you need to add `suppress_warnings = ["config.cache"]` in `doc/conf.py`. There is also an issue with the `postgis_raster` extension that can't be created during setup, I don't why.

File: .github/workflows/test_and_publish.yml [hunk 3]
```diff
@@ -84,10 +84,13 @@ jobs:
...
         # Add PostGIS extension to "gis" database
-        psql -h localhost -p 5432 -U gis -d gis -c 'CREATE EXTENSION IF NOT EXISTS postgis;'
+        psql -h localhost -p 5432 -U gis -d gis -c 'CREATE EXTENSION IF NOT EXISTS postgis SCHEMA public;'
+
+        # Add PostGISRaster extension to "gis" database
+        psql -h localhost -p 5432 -U gis -d gis -c 'CREATE EXTENSION IF NOT EXISTS postgis_raster SCHEMA public;'
 
         # Drop PostGIS Tiger Geocoder extension to "gis" database
-        psql -h localhost -p 5432 -U gis -d gis -c 'DROP EXTENSION IF EXISTS postgis_tiger_geocoder CASCADE;'
+        psql -h localhost -p 5432 -U gis -d gis -c 'DROP EXTENSION IF EXISTS postgis_tiger_geocoder SCHEMA public CASCADE;'
```
Comment (adrien-berchet) [inline .github/workflows/test_and_publish.yml:93]:
```suggestion
        psql -h localhost -p 5432 -U gis -d gis -c 'DROP EXTENSION IF EXISTS postgis_tiger_geocoder CASCADE;'
```
Ups, sorry for this one :smile:

File: .github/workflows/test_and_publish.yml [hunk 4]
```diff
@@ -84,7 +84,10 @@ jobs:
...
 
         # Add PostGIS extension to "gis" database
-        psql -h localhost -p 5432 -U gis -d gis -c 'CREATE EXTENSION IF NOT EXISTS postgis;'
+        psql -h localhost -p 5432 -U gis -d gis -c 'CREATE EXTENSION IF NOT EXISTS postgis SCHEMA public;'
+
+        # Add PostGISRaster extension to "gis" database
+        psql -h localhost -p 5432 -U gis -d gis -c 'CREATE EXTENSION IF NOT EXISTS postgis_raster SCHEMA public;'
 
         # Drop PostGIS Tiger Geocoder extension to "gis" database
         psql -h localhost -p 5432 -U gis -d gis -c 'DROP EXTENSION IF EXISTS postgis_tiger_geocoder CASCADE;'
```
Comment (adrien-berchet) [inline .github/workflows/test_and_publish.yml:93]:
```suggestion
        # Drop PostGIS Topology extension to "gis" database
        psql -h localhost -p 5432 -U gis -d gis -c 'DROP EXTENSION IF EXISTS postgis_topology;'
```

File: .github/workflows/test_and_publish.yml [hunk 5]
```diff
@@ -84,10 +84,13 @@ jobs:
         psql -h localhost -p 5432 -U gis -d gis -c 'CREATE SCHEMA gis;'
 
         # Add PostGIS extension to "gis" database
-        psql -h localhost -p 5432 -U gis -d gis -c 'CREATE EXTENSION IF NOT EXISTS postgis;'
+        psql -h localhost -p 5432 -U gis -d gis -c 'CREATE EXTENSION IF NOT EXISTS postgis SCHEMA public;'
 
-        # Drop PostGIS Tiger Geocoder extension to "gis" database
-        psql -h localhost -p 5432 -U gis -d gis -c 'DROP EXTENSION IF EXISTS postgis_tiger_geocoder CASCADE;'
+        # Add PostGISRaster extension to "gis" database
```
Comment (adrien-berchet) [inline .github/workflows/test_and_publish.yml:89]:
```suggestion
        # Drop PostGIS Tiger Geocoder extension to "gis" database
        psql -h localhost -p 5432 -U gis -d gis -c 'DROP EXTENSION IF EXISTS postgis_tiger_geocoder CASCADE;'

        # Add PostGISRaster extension to "gis" database
```
Seems it's actually still needed...

Comment (sdp5) [reply .github/workflows/test_and_publish.yml:92]: okay

**Comment by adrien-berchet** [comment]
Looks good this time :partying_face: 
Do you think it's ok to be merged now or do you still have some tests to do @sdp5 ?

**Comment by sdp5** [comment]
🎉 thanks, please go ahead merging @adrien-berchet !

**Comment by adrien-berchet** [comment]
Thank you very much for this work @sdp5 !!!

**Pull request merged**